File: backend/src/main/java/lan/dk/podcastserver/manager/worker/francetv/FranceTvExtractor.java
package lan.dk.podcastserver.manager.worker.francetv;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.davinkevin.podcastserver.manager.worker.francetv.FranceTvUpdater;
import com.github.davinkevin.podcastserver.service.HtmlService;
import io.vavr.collection.List;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.manager.downloader.DownloadingItem;
import lan.dk.podcastserver.manager.worker.Extractor;
import lan.dk.podcastserver.service.JsonService;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import static io.vavr.API.*;

/**
 * Created by kevin on 24/12/2017
 */
@Component
@Scope("prototype")
@RequiredArgsConstructor
public class FranceTvExtractor implements Extractor {

    private static final String CATALOG_URL = "https://sivideo.webservices.francetelevisions.fr/tools/getInfosOeuvre/v2/?idDiffusion=%s";

    private final HtmlService htmlService;
    private final JsonService jsonService;

    @Override
    public DownloadingItem extract(Item item) {
        return htmlService.get(item.getUrl())
                .map(d -> d.select("#player").first())
                .map(e -> e.attr("data-main-video"))
                .flatMap(id -> jsonService.parseUrl(String.format(CATALOG_URL, id)))
                .map(JsonService.to(FranceTvItem.class))
                .map(FranceTvItem::getUrl)
                .map(url -> Tuple(url, getFileName(item)))
                .map(info -> DownloadingItem.builder()
                        .filename(info._2())
                        .urls(List(info._1()))
                        .item(item)
                    .build())
                .getOrElseThrow(() -> new RuntimeException("Error during extraction of FranceTV item"));
    }


    @Override
    public Integer compatibility(String url) {
        return FranceTvUpdater.Companion.isFromFranceTv(url);
    }

    @Override
    public String getFileName(Item item) {
        return Option(item.getUrl())
                .map(s -> StringUtils.substringAfterLast(s, "/"))
                .map(s -> StringUtils.substringBeforeLast(s, "?"))
                .map(FilenameUtils::getBaseName)
                .map(s -> s + ".mp4")
                .getOrElse("");
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class FranceTvItem {

        @Setter @Getter private List<FranceTvItem.Video> videos = List.empty();

        public String getUrl() {
            return videos
                    .find(v -> "hls_v5_os".equals(v.getFormat()))
                    .orElse(videos.find(v -> "m3u8-download".equals(v.getFormat())))
                    .orElse(videos.find(v -> v.getSecureUrl().contains("m3u8")))
                    .flatMap(v -> Option(v.getSecureUrl()).orElse(Option(v.getUrl())))
                    .getOrElseThrow(() -> new RuntimeException("No video found in this FranceTvItem " + this));

        }

        @JsonIgnoreProperties(ignoreUnknown = true)
        private static class Video {
            @Getter @Setter private String format;
            @Getter @Setter private String url;
            @JsonProperty("url_secure") @Getter @Setter private String secureUrl;
        }
    }


}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/francetv/FranceTvExtractor.kt
package com.github.davinkevin.podcastserver.manager.worker.francetv

import arrow.core.getOrElse
import arrow.core.orElse
import arrow.core.toOption
import arrow.syntax.collections.firstOption
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.davinkevin.podcastserver.entity.Item
import com.github.davinkevin.podcastserver.manager.downloader.DownloadingItem
import com.github.davinkevin.podcastserver.manager.worker.Extractor
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.utils.k
import lan.dk.podcastserver.service.JsonService
import org.apache.commons.io.FilenameUtils
import org.jsoup.select.Elements
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
 * Created by kevin on 24/12/2017
 */
@Component
@Scope(SCOPE_PROTOTYPE)
class FranceTvExtractor(val htmlService: HtmlService, val jsonService: JsonService) : Extractor {

    private val log = LoggerFactory.getLogger(FranceTvExtractor::class.java)!!

    override fun extract(item: Item) =
            htmlService
                    .get(item.url!!)
                    .map { it.select("script") }
                    .getOrElse {
                        log.error("No script found for item ${item.url!!}")
                        Elements()
                    }
                    .map { it.html() }
                    .firstOption { it.contains("FTVPlayerVideos") }
                    .map { it.substringAfter("=").trim(';') }
                    .flatMap { jsonService.parse(it).toOption() }
                    .map { JsonService.to(FranceTvUpdater.PAGE_ITEM).apply(it) }
                    .getOrElse { setOf() }
                    .firstOption() { it.contentId in item.url!! }
                    .map { FranceTvUpdater.CATALOG_URL.format(it.videoId) }
                    .flatMap {
                        log.info("parsing url catalog $it")
                        val parseUrl = jsonService.parseUrl(it)
                        parseUrl.k()
                    }
                    .map { JsonService.to(FranceTvVideoItem::class.java).apply(it) }
                    .map { it.url }
                    .map { DownloadingItem(item, listOf(it), item.url!!.filename().baseName() + ".mp4", null) }
                    .getOrElse { throw RuntimeException("Error during extraction of FranceTV item") }

    override fun compatibility(url: String?) = FranceTvUpdater.isFromFranceTv(url)
}

@JsonIgnoreProperties(ignoreUnknown = true)
private class FranceTvVideoItem {

    var videos = listOf<Video>()

    val url: String
        get() {
            val onlineVideos = videos.filter { it.statut != "OFFLINE" }
            return onlineVideos
                    .firstOption { "hls_v5_os" == it.format }
                    .orElse { onlineVideos.firstOption { "m3u8-download" == it.format } }
                    .orElse { onlineVideos.firstOption { (it.secureUrl ?: "").contains("m3u8") }}
                    .map { it.secureUrl ?: it.url }
                    .getOrElse { throw RuntimeException("No video found in this FranceTvItem ") }!!
        }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private class Video {
        var format: String? = null
        var url: String? = null
        var statut: String? = null
        @JsonProperty("url_secure") var secureUrl: String? = null
    }
}

private fun String.filename() = this.substringAfterLast("/").substringBeforeLast("?")
private fun String.baseName() = FilenameUtils.getBaseName(this)
